core/src/main/java/org/jboss/set/mjolnir/archive/batch/RemoveArchieveUserBatchlet.java

```diff
@@ -0,0 +1,96 @@
+package org.jboss.set.mjolnir.archive.batch;
+
+import org.jboss.logging.Logger;
+import org.jboss.set.mjolnir.archive.GitDeleteRepository;
+import org.jboss.set.mjolnir.archive.configuration.Configuration;
+import org.jboss.set.mjolnir.archive.domain.RepositoryFork;
+
+import javax.batch.api.AbstractBatchlet;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.persistence.EntityManager;
+import javax.persistence.EntityTransaction;
+import java.sql.Timestamp;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+@Named
+public class RemoveArchieveUserBatchlet extends AbstractBatchlet {
+
+    private final Logger logger = Logger.getLogger(getClass());
+
+    @Inject
+    private EntityManager em;
+
+    @Inject
+    private Configuration configuration;
+
+    @Override
+    public String process() throws Exception {
+
+        logger.infof("User removal process started");
+        boolean successful = false;
+        if (configuration.getRemoveArchives() == true) {
+            List<RepositoryFork> removals = loadRepositoryForks();
+            List<RepositoryFork> filteredList = getFilterDeleteList(removals);
+            for (RepositoryFork repositoryFork : filteredList) {
+                GitDeleteRepository gitDeleteRepository = new GitDeleteRepository(configuration.getRepositoryArchiveRoot() + "/" + repositoryFork.getSourceRepositoryName(),
+                        repositoryFork.getRepositoryName().substring(0, repositoryFork.getRepositoryName().indexOf('/')));
+
+                gitDeleteRepository.gitRemoveBranches();
+                successful = true;
+            }
+
+            if (successful) {
+                return "DONE";
+            } else {
+                return "DONE_WITH_ERRORS";
+            }
+        } else {
+            return "APPLICATION_REMOVE_ARCHIEVE_NOT_APPLICABLE";
+        }
+    }
+
+    public List<RepositoryFork> loadRepositoryForks() {
+        // perform in transaction to avoid removals being loaded by two parallel executions
+        HashMap<String, String> map = new HashMap<String, String>();
+        EntityTransaction transaction = em.getTransaction();
+        transaction.begin();
+        List<RepositoryFork> list = new ArrayList<RepositoryFork>();
+        List<Object[]> removals = em.createNativeQuery("SELECT r.source_repository_name, r.repository_name, MAX(r.created) as created FROM REPOSITORY_FORKS r " +
+                "GROUP BY r.source_repository_name, r.repository_name")
+                .getResultList();
```

Comment:
We could move this query to `RepositoryFork` class - make it a *named query*, similarly like we have named queries in `RegisteredUser` class. That way the JPQL will always be in entity classes, easy to locate...

Reply:
Also, I think it would be neater to move the time check into the JPQL query, rather than do it in Java. The query could be something like:

```
SELECT r FROM RepositoryFork r
WHERE r.deleted is null
HAVING max(r.created) + interval :age day < CURRENT_TIMESTAMP
```
but I haved tried to run it :).

You can then call it with something like

```
List<RegisteredUser> resultList = em.createNamedQuery(RepositoryFork.FORKS_TO_REMOVE, RepositoryFork.class)
    .setParameter("age", configuration.getRemoveArchivesAfter())
    .getResultList();
```

and you would get `RepositoryFork` objects directly, instead of record arrays.

Reply:
Tried above query, not working and lots of other issues. instead selected all where deleted is null and rest logic in java code.

Reply:
OK, I'm checking this out, it indeed looks a bit more difficult...